src/matrix/verification/SAS/channel/MockChannel.ts

```diff
@@ -0,0 +1,139 @@
+            this.receivedMessages.set(eventType, startingMessage);
+        }
+    }
+
+    async send(eventType: string, content: any, _: ILogItem) {
+        if (this.isCancelled) {
+            throw new VerificationCancelledError();
+        }
+        Object.assign(content, { transaction_id: this.id });
+        this.sentMessages.set(eventType, {content});
+    }
+
+    async waitForEvent(eventType: string): Promise<any> {
+        if (this.isCancelled) {
+            throw new VerificationCancelledError();
+        }
+        const event = this.fixtures.get(eventType);
+        if (event) {
+            this.receivedMessages.set(eventType, event);
+        }
+        else {
+            await new Promise(() => {});
+        }
+        if (eventType === VerificationEventTypes.Mac) {
+            await this.recalculateMAC();
+        }
+        if(eventType === VerificationEventTypes.Accept && this.startMessage) {
+        }
+        return event;
+    }
+
+    private recalculateCommitment() {
+        const acceptMessage = this.acceptMessage?.content;
+        if (!acceptMessage) {
+            return;
+        }
+        const {content} = this.startMessage;
+        const {content: keyMessage} = this.fixtures.get(VerificationEventTypes.Key);
+        const key = keyMessage.key;
+        const commitmentStr = key + anotherjson.stringify(content);
+        const olmUtil = new this.olm.Utility();
+        const commitment = olmUtil.sha256(commitmentStr);
+        olmUtil.free();
+        acceptMessage.commitment = commitment;
+    }
+
+    private async recalculateMAC() {
+        // We need to replace the mac with calculated mac
+        const baseInfo =
+            "MATRIX_KEY_VERIFICATION_MAC" +
+            this.otherUserId +
+            this.otherUserDeviceId +
+            this.ourUserId +
+            this.ourUserDeviceId +
+            this.id;
+        const { content: macContent } = this.receivedMessages.get(VerificationEventTypes.Mac);
+        const macMethod = this.acceptMessage.content.message_authentication_code;
+        const calculateMac = createCalculateMAC(this.olmSas, macMethod);
+        const input = Object.keys(macContent.mac).sort().join(",");
+        const properMac = calculateMac(input, baseInfo + "KEY_IDS");
+        macContent.keys = properMac;
+        for (const keyId of Object.keys(macContent.mac)) {
+            const deviceId = keyId.split(":", 2)[1];
+            const device = await this.deviceTracker.deviceForId(this.otherUserDeviceId, deviceId);
+            if (device) {
+                macContent.mac[keyId] = calculateMac(device.ed25519Key, baseInfo + keyId);
+            }
+            else {
+                const {masterKey} = await this.deviceTracker.getCrossSigningKeysForUser(this.otherUserId);
+                macContent.mac[keyId] = calculateMac(masterKey, baseInfo + keyId);
+            }
+        }
+    }
+
+    setStartMessage(event: any): void {
```

Comment:
Yes, because in the case where you send the start message but also receive a start message, there's a conflict. 
This is resolved in `SelectVerificationMethodStage` which tells the channel which the "true" start message is.

Reply:
aha, that makes sense, thanks!